src/libsyntax/ext/tt/macro_rules.rs

```diff
@@ -253,7 +254,8 @@ pub fn compile(sess: &ParseSess, features: &RefCell<Features>, def: &ast::Item)
             s.iter().map(|m| {
                 if let MatchedNonterminal(ref nt) = *m {
                     if let NtTT(ref tt) = **nt {
-                        return quoted::parse(tt.clone().into(), false, sess).pop().unwrap();
+                        return quoted::parse(tt.clone().into(), !body.legacy, false, sess)
```

Comment:
Could you add two test cases for `#` inside of `macro_rules` inside of `macro`, and for `#` inside of `macro` inside of `macro_rules`.

Reply:
Just to be clear: how do want this to work?

Reply:
`#` is a separate token inside of `macro_rules` regardless of outer context, `#` is a hygiene modifier in `macro` regardless of outer context. I think this is exactly what the current implementation does.

Reply:
https://github.com/rust-lang/rust/pull/47992#issuecomment-371081641